### PR TITLE
Update to decodeAudioData() docs page

### PR DESCRIPTION
**Pull request by sidewayss**
The current page for decodeAudioData() is bad at explaining the newer promise-based syntax, and it is the newer, better way to go. These edits update the page to reflect the newer reality of promises and async/await.  The XHR/callback syntax example has also been updated and this PR reflects those changes too.
This PR replaces #26839, which is now closed.
This PR aligns with this web-audio-examples PR: https://github.com/mdn/webaudio-examples/pull/109

I was unable to create github pages for this repo, which makes sense since it's so large.  Now I see that the PR creates a preview page automatically.  I'll try to remember that for next time...

### CODE REVIEW

**Comment by sidewayss** [comment]
I removed all the example code that doesn't directly relate to `decodeAudioData`.  The current page's XHR example code includes a whole lot more.  IMO it's best this way, and people can go to the live examples to see the rest of it, but it might make sense to include a stripped down version of the play button code either in both examples (play button code is identical in both examples) or as a separate example section.  Let me know what you think about that.

File: files/en-us/web/api/baseaudiocontext/decodeaudiodata/index.md [hunk 1]
```diff
@@ -48,97 +47,94 @@ decodeAudioData(arrayBuffer)
 
 ### Return value
 
-None ({{jsxref("undefined")}}) or a {{jsxref("Promise") }} object that fulfills with the
-_decodedData_.
+If a callback is used, ({{jsxref("undefined")}}), else a {{jsxref("Promise") }} object that
+fulfills with the _decodedData_.
```
Comment (wbamberg) [inline files/en-us/web/api/baseaudiocontext/decodeaudiodata/index.md:51]: I looks like this function always returns a Promise: https://webaudio.github.io/web-audio-api/#dom-baseaudiocontext-decodeaudiodata ?

File: files/en-us/web/api/baseaudiocontext/decodeaudiodata/index.md [hunk 2]
```diff
@@ -48,97 +47,94 @@ decodeAudioData(arrayBuffer)
...
-setting the `responseType` of the request to `arraybuffer` so that
-it returns an array buffer as its `response` that we then store in the
-`audioData` variable. We then pass this buffer into a
-`decodeAudioData()` function; the success callback takes the successfully
-decoded PCM data, puts it into an {{ domxref("AudioBufferSourceNode") }} created using
-{{domxref("BaseAudioContext/createBufferSource", "AudioContext.createBufferSource()")}}, connects the source to the
-{{domxref("BaseAudioContext/destination", "AudioContext.destination")}} and sets it to loop.
+In this example `fetchData()` uses {{domxref("fetch()")}} to retrieve an audio
+file asynchronously and decodes it into an `audioBuffer`. It then caches the
+`audioBuffer` in the global `buffer variable` for later playback.
```
Comment (wbamberg) [inline files/en-us/web/api/baseaudiocontext/decodeaudiodata/index.md:61]:
```suggestion
`audioBuffer` in the global `buffer` variable for later playback.
```

**Comment by sidewayss** [comment]
What just happened here?  This PR has been idle for weeks.  Did another PR change the page?

**Comment by wbamberg** [comment]
> What just happened here? This PR has been idle for weeks. Did another PR change the page?

Yes, we upgraded to Prettier 3 and it caused a lot of churn: https://github.com/mdn/content/pull/27777.

There is still an outstanding comment on this PR: https://github.com/mdn/content/pull/27185#pullrequestreview-1465922967 , are you planning to come back to it.

**Comment by sidewayss** [comment]
@wbamberg - I thought the ball was in your court.  I replied to your review comment extensively and was waiting for a response from you (see above).  How do you want to handle this?  I suggested simply ignoring it because functionally it makes no difference and documenting it could cause confusion.  But if you want to mention it I can make changes to do so.
Do I need to rebase to make this mergeable again?  Or is the fix to run Prettier again?

**Comment by wbamberg** [comment]
> @wbamberg - I thought the ball was in your court. I replied to your review comment extensively and was waiting for a response from you (see above). How do you want to handle this? I suggested simply ignoring it because functionally it makes no difference and documenting it could cause confusion. But if you want to mention it I can make changes to do so. Do I need to rebase to make this mergeable again? Or is the fix to run Prettier again?

I don't see any response to https://github.com/mdn/content/pull/27185#pullrequestreview-1465922967 here, can you link to it please?

>  I suggested simply ignoring it because functionally it makes no difference and documenting it could cause confusion.

If this refers to https://github.com/mdn/content/pull/27185#pullrequestreview-1465922967, then I think it matters because it's incorrect to say that the function returns `undefined` if a callback is given.

**Comment by sidewayss** [comment]
OK, I guess I had been away from github for too long.  I didn't realize that I could commit my own review without outside approval from someone like you.  I just committed my suggested changes, but you need to let me know if you want to say anything different about the return value.  I'm in the middle of something else right now, but later I'll upgrade myself to Prettier 3.0, run this thing through it and see if that allows merging.  Unless you tell me that there is a different route for fixing it.

**Comment by sidewayss** [comment]
> I don't see any response to [#27185 (review)](https://github.com/mdn/content/pull/27185#pullrequestreview-1465922967) here, can you link to it please?
Here are my three replies:
https://github.com/mdn/content/pull/27185#discussion_r1220206962
https://github.com/mdn/content/pull/27185#discussion_r1220224584
https://github.com/mdn/content/pull/27185#discussion_r1220277307
> > I suggested simply ignoring it because functionally it makes no difference and documenting it could cause confusion.
> 
> If this refers to [#27185 (review)](https://github.com/mdn/content/pull/27185#pullrequestreview-1465922967), then I think it matters because it's incorrect to say that the function returns `undefined` if a callback is given.
It's incorrect but confusing.  Either you're ignoring the return value (XHR) or you're using it (Promise).  From a functional point of view it's two different return values.
But you're the boss, so I'll take a crack at wording it so it's as unconfusing as possible.

As for eliminating "the conflicts that must be resolved", is running Prettier 3.0 the solution?

**Comment by wbamberg** [comment]
> > I don't see any response to [#27185 (review)](https://github.com/mdn/content/pull/27185#pullrequestreview-1465922967) here, can you link to it please?
> 
> Here are my three replies: [#27185 (comment)](https://github.com/mdn/content/pull/27185#discussion_r1220206962) [#27185 (comment)](https://github.com/mdn/content/pull/27185#discussion_r1220224584) [#27185 (comment)](https://github.com/mdn/content/pull/27185#discussion_r1220277307)
> 
> > > I suggested simply ignoring it because functionally it makes no difference and documenting it could cause confusion.
> > 
> > 
> > If this refers to [#27185 (review)](https://github.com/mdn/content/pull/27185#pullrequestreview-1465922967), then I think it matters because it's incorrect to say that the function returns `undefined` if a callback is given.
> 
> It's incorrect but clear vs correct and confusing.

Being correct, in terms of stating the facts about the API, is the #1 job of reference documentation. After that, we want to be as clear as we can. I would just say it returns a promise here.

> Either you're ignoring the return value (XHR) or you're using it (Promise). 
>
> From a functional point of view it's two different return values. But you're the boss, so I'll take a crack at wording it so it's as unconfusing as possible.
>

What I think might help would be to add a paragraph after "This is the preferred method of creating an audio source for Web Audio API from an audio track. This method only works on complete file data, not fragments of audio file data.", something like:

> This function implements two alternative ways to asynchronously return the audio data or error messages: it returns a  {{jsxref("Promise")}} that fulfills with the audio data, and also accepts callback arguments to handle success or failure. The primary method of interfacing with this function is via its promise return value, and the callback parameters are provided for legacy reasons.
 
> As for eliminating "the conflicts that must be resolved", is running Prettier 3.0 the solution?

I resolved the conflicts but Prettier is still complaining. If you update your environment by running `yarn` in your `content` directory, then that should upgrade your local Prettier, and if you then make a commit, the pre-commit hooks should run Prettier for you.

**Comment by sidewayss** [comment]
I was just about to commit my changes to the return value section when I saw that you had posted here.  I had already upgraded Prettier locally.  I was prompted to do so when I tried to check my local version, so I didn't use yarn (which I've never used and am unclear as to exactly what it is).  I've added your suggested text in addition to changing the return value section's text.  I just committed and pushed, so we'll see how it goes...

**Comment by sidewayss** [comment]
Markdownlint failed again. I don't think it's running Prettier as part of the automated commit process locally.  Hopefully this will do it.  Prettier made two tiny changes to remove a comma.
And yes Markdownlint passed just now.  Good news!

**Review by wbamberg — APPROVED** [review]
Yes, looks good now 🎉 .

Thanks for all your work on this, @sidewayss !

**Comment by sidewayss** [comment]
You're welcome.  Glad to do some good.
So is running yarn the way to make git automatically run Prettier for this project?  I might contribute again, and it's good to learn about this stuff.
Are all the PRs relating to the examples merged?  I don't remember seeing any notifications about the last Promise-based example merging, but maybe github failed to notify me.  I don't have any pending PRs so they must have been merged.

**Comment by wbamberg** [comment]
> Are all the PRs relating to the examples merged? 

Yeah, this one: https://github.com/mdn/webaudio-examples/pull/107 I think.

`yarn` is just a different Node package manager, which some people, including the original author of the MDN content infrastructure, like to use instead of `npm`. So if you want to do pretty much anything with mdn/content, you're very likely to be running it: https://github.com/mdn/content/blob/main/CONTRIBUTING.md#contributing-to-mdn

So when you run `yarn`, it installs all the packages you need (including Prettier), or updates any that need updating (assuming you also have a fresh version of the mdn/content repo itself of course).

Just like `npm`, it uses `package.json` to figure out what to install (which packages, and which versions of them).

The thing is that a couple of weeks ago mdn/content updated to Prettier 3, which is a major-version update that's incompatible with 2.x (mainly because it adds trailing commas). So if you had an old version of mdn/content on your machine, then even if you were running Prettier locally, it would be using Prettier 2.8. Then you push that up to GitHub, and we run Prettier 3 on your PR, and it will complain.

This was exactly what happened to me last week. So I did something like:

```
git fetch upstream
// upstream is my name for https://github.com/mdn/content    
       
git merge upstream/main
// merge from latest mdn/content - this means we will get Prettier 3 listed in package.json

yarn
// update my local environment according to the new package.json
```

...and then it worked.

**Comment by sidewayss** [comment]
Thanks for the info on yarn.  I was just reading about it.  I was forced to install and use pnpm recently, so I'm getting plugged into all three package managers now.  I had already upgraded to Prettier 3.0, but I have never run yarn or installed any packages relating to this project, so it took running Prettier manually to handle those pesky trailing commas.
As for the Promise-base example, it looks like I never pushed my latest changes to align it with the updated XHR example.  I was sure that I had created the new branch and created the PR.  Yes, I have the branch with the latest edits.  Hmmm.  I'll confirm my latest edits and create a new PR now.

**Comment by sidewayss** [comment]
Yea, here is my PR: https://github.com/mdn/webaudio-examples/pull/109
Yet it does not appear in my list of PRs here on github.com.  You should merge this one.

**Pull request merged**